src/network/protocol/messages/getdata.zig

```diff
@@ -0,0 +1,162 @@
+        var hasher = Sha256.init(.{});
+        const writer = hasher.writer();
+        self.serializeToWriter(writer) catch unreachable; // Sha256.write is infaible
+        hasher.final(&digest);
+
+        Sha256.hash(&digest, &digest, .{});
+
+        return digest[0..4].*;
+    }
+
+    /// Free the `inventory`
+    pub fn deinit(self: GetdataMessage, allocator: std.mem.Allocator) void {
+        allocator.free(self.inventory);
+    }
+
+    /// Serialize the message as bytes and write them to the Writer.
+    ///
+    /// `w` should be a valid `Writer`.
+    pub fn serializeToWriter(self: *const GetdataMessage, w: anytype) !void {
+        comptime {
+            if (!std.meta.hasFn(@TypeOf(w), "writeInt")) @compileError("Expects writer to have fn 'writeInt'.");
+            if (!std.meta.hasFn(@TypeOf(w), "writeAll")) @compileError("Expects writer to have fn 'writeAll'.");
+        }
+
+        const count = CompactSizeUint.new(self.inventory.len);
+        try count.encodeToWriter(w);
+
+        for (self.inventory) |item| {
+            try w.writeInt(u32, item.type, .little);
+
+            try w.writeAll(&item.hash);
+        }
+    }
+
+    pub fn serialize(self: *const GetdataMessage, allocator: std.mem.Allocator) ![]u8 {
+        const serialized_len = self.hintSerializedLen();
+
+        const ret = try allocator.alloc(u8, serialized_len);
+        errdefer allocator.free(ret);
+
+        try self.serializeToSlice(ret);
+
+        return ret;
+    }
+
+    /// Serialize a message as bytes and write them to the buffer.
+   ///
+   /// buffer.len must be >= than self.hintSerializedLen()
+    pub fn serializeToSlice(self: *const GetdataMessage, buffer: []u8) !void {
+        var fbs = std.io.fixedBufferStream(buffer);
+        const writer = fbs.writer();
+        try self.serializeToWriter(writer);
+    }
+
+    pub fn hintSerializedLen(self: *const GetdataMessage) usize {
+        var length: usize = 0;
+
+        // Adding the length of CompactSizeUint for the count
+        const count = CompactSizeUint.new(self.inventory.len);
+        length += count.hint_encoded_len();
+
+        // Adding the length of each inventory item
+        length += self.inventory.len * (4 + 32); // Type (4 bytes) + Hash (32 bytes)
+
+        return length;
+    }
+
+    pub fn deserializeReader(allocator: std.mem.Allocator, r: anytype) !GetdataMessage {
+        comptime {
+            if (!std.meta.hasFn(@TypeOf(r), "readInt")) @compileError("Expects reader to have fn 'readInt'.");
+            if (!std.meta.hasFn(@TypeOf(r), "readNoEof")) @compileError("Expects reader to have fn 'readNoEof'.");
+        }
+
+        const compact_count = try CompactSizeUint.decodeReader(r);
+        const count = compact_count.value();
+
+        const inventory = try allocator.alloc(message.InventoryItem, count);
+
+        for (inventory) |*item| {
+            item.type = try r.readInt(u32, .little);
+            try r.readNoEof(&item.hash);
+        }
```

Comment:
Looks like you could impl deserializeReader serializeWriter directly onto InventoryItem

Reply:
ok